core/src/main/scala/kafka/server/ReplicaManager.scala

```diff
@@ -345,14 +365,114 @@ class ReplicaManager(val config: KafkaConfig,
+
+          val convertedOffset =
+            if (requestedOffset == PurgeRequest.HIGH_WATERMARK) {
+              partition.leaderReplicaIfLocal match {
+                case Some(leaderReplica) =>
+                  leaderReplica.highWatermark.messageOffset
+                case None =>
+                  throw new NotLeaderForPartitionException("Leader not local for partition %s on broker %d"
+                    .format(topicPartition, localBrokerId))
+              }
+            } else
+              requestedOffset
+          if (convertedOffset < 0)
+            throw new OffsetOutOfRangeException(s"The purge offset $convertedOffset for partition $topicPartition is not valid")
+
+          val lowWatermark = partition.purgeRecordsOnLeader(convertedOffset)
+          (topicPartition, LogPurgeResult(convertedOffset, lowWatermark))
+        } catch {
+          // NOTE: Failed produce requests metric is not incremented for known exceptions
+          // it is supposed to indicate un-expected failures of a broker in handling a produce request
+          case e: KafkaStorageException =>
+            fatal("Halting due to unrecoverable I/O error while handling purge request: ", e)
+            Runtime.getRuntime.halt(1)
+            (topicPartition, null)
+          case e@ (_: UnknownTopicOrPartitionException |
+                   _: NotLeaderForPartitionException |
+                   _: OffsetOutOfRangeException |
+                   _: NotEnoughReplicasException) =>
+            (topicPartition, LogPurgeResult(-1L, -1L, Some(e)))
+          case t: Throwable =>
+            error("Error processing purge operation on partition %s".format(topicPartition), t)
+            (topicPartition, LogPurgeResult(-1L, -1L, Some(t)))
+        }
+      }
+    }
+  }
+
+  // If there exists a topic partition that meets the following requirement,
+  // we need to put a delayed purge request and wait for the purge operation to complete
+  //
+  // 1. the purge operation on this partition is successful
+  // 2. low watermark of this partition is smaller than the purgeOffset
+  private def delayedPurgeRequestRequired(localPurgeResults: Map[TopicPartition, LogPurgeResult]): Boolean = {
+    localPurgeResults.exists{ case (tp, purgeResult) =>
+      purgeResult.exception.isEmpty && purgeResult.lowWatermark < purgeResult.requestedOffset
+    }
+  }
+
+  def purgeRecords(timeout: Long,
+                   offsetPerPartition: Map[TopicPartition, Long],
+                   metadataCache: MetadataCache,
+                   responseCallback: Map[TopicPartition, PurgeResponse.PartitionResponse] => Unit) {
+    val sTime = if (logger.isDebugEnabled) time.milliseconds else 0
+    val localPurgeResults = purgeOnLocalLog(offsetPerPartition, metadataCache)
+    debug("Purge on local log in %d ms".format(time.milliseconds - sTime))
+
+    val purgeStatus = localPurgeResults.map { case (topicPartition, result) =>
```

Comment:
I validated that we can not use `mapValues()` here. This is because `mapValues()` returns a map view which maps every key of this map to `f(this(key))`. The resulting map wraps the original map without copying any elements. As a result `status.acksPending = true` in the constructor of `DelayedPurge` becomes no-op.